docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/EventEmitter.java

```diff
@@ -0,0 +1,134 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.docker.compose;
+
+import com.google.common.base.Throwables;
+import com.palantir.docker.compose.connection.Cluster;
+import com.palantir.docker.compose.connection.waiting.ClusterWait;
+import com.palantir.docker.compose.events.BuildEvent;
+import com.palantir.docker.compose.events.ClusterWaitEvent;
+import com.palantir.docker.compose.events.ClusterWaitEvent.ClusterWaitType;
+import com.palantir.docker.compose.events.DockerComposeRuleEvent;
+import com.palantir.docker.compose.events.EventConsumer;
+import com.palantir.docker.compose.events.PullImagesEvent;
+import com.palantir.docker.compose.events.ShutdownEvent;
+import com.palantir.docker.compose.events.TaskEvent;
+import com.palantir.docker.compose.events.UpEvent;
+import com.palantir.docker.compose.events.WaitForServicesEvent;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class EventEmitter {
+    private static final Logger log = LoggerFactory.getLogger(EventEmitter.class);
+
+    private List<EventConsumer> eventConsumers;
+
+    EventEmitter(List<EventConsumer> eventConsumers) {
+        this.eventConsumers = eventConsumers;
+    }
+
+    interface CheckedRunnable {
+        void run() throws InterruptedException, IOException;
+    }
+
+    public void pull(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, PullImagesEvent.FACTORY);
+    }
+
+    public void build(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, BuildEvent.FACTORY);
+    }
+
+    public void up(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, UpEvent.FACTORY);
+    }
+
+    public void waitingForServices(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, WaitForServicesEvent.FACTORY);
+    }
+
+    public void shutdown(CheckedRunnable runnable) throws IOException, InterruptedException {
+        emitThrowing(runnable, ShutdownEvent.FACTORY);
+    }
+
+    public Consumer<Cluster> userClusterWait(ClusterWait clusterWait)  {
+        return clusterWait(ClusterWaitType.USER, clusterWait);
+    }
+
+    public Consumer<Cluster> nativeClusterWait(ClusterWait clusterWait) {
+        return clusterWait(ClusterWaitType.NATIVE, clusterWait);
+    }
+
+    private Consumer<Cluster> clusterWait(
```

Comment:
gotta justify why we have to do this - clusterwaits don't declare which containers they access, so we have to figure it out while they are executing??